### ISSUE TITLE
lit-element props configured by TS decorators do not cause a re-render of the component

### ISSUE THREAD

**Issue by steveworkman** [storybookjs/storybook#12369]
**Describe the bug**
With Lit-element, using controls to update the element's properties, storybook does not cause a render of the component when the component's properties are defined by TypeScript decorators.

**To Reproduce**
https://github.com/steveworkman/sb6-ts-decorators-bug

1. In that repo - run the storybook
2. In the non-decorators story, change the "kind" to be "secondary" - see that the component re-renders and the colour changtes
3. Do the same in the decorators story

**Expected behavior**
The component should re-render the same as the non-decorator one

**Code snippets**
Repo: https://github.com/steveworkman/sb6-ts-decorators-bug

**System:**
```
Environment Info:

  System:
    OS: macOS 10.15.6
    CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  Binaries:
    Node: 12.16.1 - ~/.nvm/versions/node/v12.16.1/bin/node
    npm: 6.14.7 - ~/.nvm/versions/node/v12.16.1/bin/npm
  Browsers:
    Chrome: 85.0.4183.83
    Safari: 13.1.2
  npmPackages:
    @storybook/addon-actions: ^6.0.21 => 6.0.21 
    @storybook/addon-essentials: ^6.0.21 => 6.0.21 
    @storybook/addon-links: ^6.0.21 => 6.0.21 
    @storybook/web-components: ^6.0.21 => 6.0.21 
```

**Additional context**
I think it's to do with how the TS decorators are transpiled, but I can't be sure. This was working in SB 5 with a webcomponents setup.

**Comment by steveworkman** [comment]
This is the compiled output of the top of the`decorators-button`

```js
var DecoratorsButton = (_dec = Object(lit_element__WEBPACK_IMPORTED_MODULE_25__["customElement"])("decorators-button"), _dec2 = Object(lit_element__WEBPACK_IMPORTED_MODULE_25__["property"])({
  type: String
}), _dec3 = Object(lit_element__WEBPACK_IMPORTED_MODULE_25__["property"])({
  type: String
}), _dec4 = Object(lit_element__WEBPACK_IMPORTED_MODULE_25__["property"])({
  type: String
}), _dec5 = Object(lit_element__WEBPACK_IMPORTED_MODULE_25__["property"])({
  type: Boolean
}), _dec6 = Object(lit_element__WEBPACK_IMPORTED_MODULE_25__["property"])({
  type: Function
}), _dec(_class = (_class2 = (_temp = /*#__PURE__*/function (_LitElement) {
  _inherits(DecoratorsButton, _LitElement);

  var _super = _createSuper(DecoratorsButton);

  function DecoratorsButton() {
    var _this;

    _classCallCheck(this, DecoratorsButton);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _initializerDefineProperty(_this, "kind", _descriptor, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "name", _descriptor2, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "type", _descriptor3, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "disabled", _descriptor4, _assertThisInitialized(_this));

    _initializerDefineProperty(_this, "onButtonMouseOver", _descriptor5, _assertThisInitialized(_this));

    return _this;
  }
```

And this is the same for the `non-decorators-button`

```js
var NonDecoratorsButton = (_dec = Object(lit_element__WEBPACK_IMPORTED_MODULE_19__["customElement"])("non-decorators-button"), _dec(_class = /*#__PURE__*/function (_LitElement) {
  _inherits(NonDecoratorsButton, _LitElement);

  var _super = _createSuper(NonDecoratorsButton);

  _createClass(NonDecoratorsButton, null, [{
    key: "styles",
    get: function get() {
      return [Object(lit_element__WEBPACK_IMPORTED_MODULE_19__["css"])(_templateObject())];
    }
  }, {
    key: "properties",
    get: function get() {
      return {
        kind: {
          type: String
        },
        name: {
          type: String
        },
        type: {
          type: String
        },
        disabled: {
          type: Boolean
        },
        onButtonMouseOver: {
          type: Function
        }
      };
    }
  }]);

  function NonDecoratorsButton() {
    var _this;

    _classCallCheck(this, NonDecoratorsButton);

    _this = _super.call(this);
    _this.kind = "primary";
    _this.name = "button";
    _this.type = "button";
    _this.disabled = false;

    _this.onButtonMouseOver = function () {};

    return _this;
  }
```

Are these two things equivalent to storybook?

**Comment by justinfagnani** [comment]
This is probably caused by the `useDefineForClassFields` option: https://github.com/microsoft/TypeScript/issues/35081

That option is a pretty big breaking change, if Storybook turns it on, it should turn it off.

**Comment by steveworkman** [comment]
I can't find that option anywhere in the source code. I think this may be more down to using `babel-loader` and https://babeljs.io/docs/en/next/babel-plugin-transform-typescript.html to do the transform 

Feels like this option is deeper down there somewhere

**Comment by MaxAltena** [comment]
Having the same issue. I fixed it with https://github.com/Polymer/lit-element/issues/1052 but that breaks normal usage, so not ideal.

**Comment by emilio-martinez** [comment]
I'm not able to fully dig in now, but this does sounds familiar. I believe this is a Babel-related issue when applying the following combination of options within the Babel config:

```js
// Storybook is likely using these options within its Babel
"plugins": [
  ["@babel/plugin-proposal-decorators", { "legacy": true }],
  ["@babel/plugin-proposal-class-properties", { "loose": true }]
]
```

TL;DR, use these options instead

```js
"plugins": [
  [ "@babel/plugin-proposal-decorators", { "decoratorsBeforeExport": true } ],
  "@babel/plugin-proposal-class-properties"
]
```

LitElement has two decorator implementations, attempting to support two different stages of decorator proposals. The Typescript compiler (`tsc`) uses the older proposal, while Babel's `@babel/plugin-proposal-decorators` by default will use the newer one. However, `@babel/plugin-proposal-decorators` optionally allows for a `legacy` flag, which instead switches to support the same proposal used with `tsc`.

However, that `legacy` option from Babel doesn't work well with LitElement decorators + Typescript. Instead, when using Babel you should target the non-legacy variant and enable the `decoratorsBeforeExport` option as shown above. I do remember `useDefineForClassFields` being an issue in some cases, as Justin mentioned, but I don't remember that being related to this particular issue. Incidentally, Babel applies that behavior via `@babel/plugin-proposal-class-properties`.

**Comment by MaxAltena** [comment]
The proposed fix by @emilio-martinez works for me. I added a `babel.config.js` into the `.storybook/` directory with the plugins. Thanks!

**Comment by himerus** [comment]
Still seeing this issue with latest Storybook + Lit2. The workaround doesn't seem to work at this point, or I'm doing something goofy.

**Comment by himerus** [comment]
Scratch that, I ended up pulling over the full babel config from storybook and modifying. 

The following adjustments are what removed errors and most warnings regarding loose mode.
```
[
        '@babel/plugin-proposal-decorators',
        {
            decoratorsBeforeExport: true,
        },
      ],
      ['@babel/plugin-proposal-class-properties', { loose: false }],
      ['@babel/plugin-proposal-private-methods', { loose: false }],
```
